### PR TITLE
Stringify, parse, copy, stringify and parsing causes duplicate contained resources

### LINKED ISSUE #4837: Possible bug in the fhir json parser
I'm seeing some strange behaviour in the fhir json parser. If I first copy a resource and then run it through the parser, then all referenced contained resources are duplicated. Is this a bug, or is it perhaps possible to configure the parser to somehow avoid this behaviour? I'm using hapi-fhir-base 6.4.4, Java 20 on MacOS Ventura 13.3.1.

Example to reproduce the problem:

```
FhirContext fhirContext = FhirContext.forR4();
IParser parser = fhirContext.newJsonParser();

AllergyIntolerance resource = new AllergyIntolerance();
resource.setRecorder(new Reference("#recorderRef"));

PractitionerRole pr = new PractitionerRole();
pr.setId("#recorderRef");
resource.addContained(pr);

String resourceString = parser.encodeResourceToString(resource);
AllergyIntolerance parsedResource = parser.parseResource(AllergyIntolerance.class, resourceString);

assert parsedResource.getContained().size() == 1; // Only 1 contained resource at this point

AllergyIntolerance copy = parsedResource.copy();

assert copy.getContained().size() == 1; // Only 1 contained resource in copy

String stringCopy = parser.encodeResourceToString(copy);
AllergyIntolerance parsedCopy = parser.parseResource(AllergyIntolerance.class, stringCopy);

assert parsedCopy.getContained().size() == 1; // This fails, now there are 2 contained resources with id "#recorderRef". Why?
```

### PR DESCRIPTION
**Pull request by TahaAttari**
fix #4837

### CODE REVIEW

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/FhirTerser.java [hunk 1]
```diff
@@ -1817,7 +1816,15 @@ public IIdType getResourceId(IBaseResource theNext) {
...
-			return getResourceToIdMap().get(theNext);
+			var idFromMap = getResourceToIdMap().get(theNext);
+			if (idFromMap != null) {
+				return idFromMap;
+			} else {
+				var checkAllIdsForStringMatch = getResourceToIdMap().values().stream()
+					.filter(id -> theNext.getIdElement().getIdPart().equals(id.getIdPart()))
+					.findAny();
+				return checkAllIdsForStringMatch.orElse(null);
+			}
```
Comment (TahaAttari) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/util/FhirTerser.java:1827]: important change

File: hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/util/FhirTerserR4Test.java [hunk 2]
```diff
@@ -1520,23 +1522,44 @@ void extensionWithEnumeration() {
...
+   private List<String> toStrings(List<StringType> theStrings) {
+	  ArrayList<String> retVal = new ArrayList<>();
+	  for (StringType next : theStrings) {
+		 retVal.add(next.getValue());
+	  }
+	  return retVal;
+   }
+
+   @Test
+   void copyingAndParsingCreatesDuplicateContainedResources() {
```
Comment (TahaAttari) [inline hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/util/FhirTerserR4Test.java:1534]: new test

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/FhirTerser.java [hunk 3]
```diff
@@ -1817,7 +1816,15 @@ public IIdType getResourceId(IBaseResource theNext) {
...
-			return getResourceToIdMap().get(theNext);
+			var idFromMap = getResourceToIdMap().get(theNext);
+			if (idFromMap != null) {
+				return idFromMap;
+			} else {
+				var checkAllIdsForStringMatch = getResourceToIdMap().values().stream()
+						.filter(id -> theNext.getIdElement().getIdPart().equals(id.getIdPart()))
+						.findAny();
+				return checkAllIdsForStringMatch.orElse(null);
+			}
```
Comment (TahaAttari) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/util/FhirTerser.java:1827]: important part

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/FhirTerser.java [hunk 4]
```diff
@@ -1817,7 +1816,15 @@ public IIdType getResourceId(IBaseResource theNext) {
 			if (getResourceToIdMap() == null) {
 				return null;
 			}
-			return getResourceToIdMap().get(theNext);
+			var idFromMap = getResourceToIdMap().get(theNext);
+			if (idFromMap != null) {
+				return idFromMap;
+			} else {
+				var checkAllIdsForStringMatch = getResourceToIdMap().values().stream()
```
Comment (jmarchionatto) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/util/FhirTerser.java:1823]: nitpick: var not needed. Can add orElse and just return from the stream

**Review by jmarchionatto — APPROVED** [review]
Looks good! Requires changelog. Approved pending changelog addition.

**Comment by TahaAttari** [comment]
already merged by https://github.com/hapifhir/hapi-fhir/pull/6177

**Pull request closed without merging**